SCons/Tool/MSCommon/vc.py

```diff
@@ -1515,18 +1515,30 @@ def msvc_setup_env(env):
         SCons.Warnings.warn(SCons.Warnings.VisualCMissingWarning, warn_msg)
         return None
 
+    found_cl_path = None
+    found_cl_envpath = None
+
+    seen_path = False
     for k, v in d.items():
+        if not seen_path and k == 'PATH':
+            seen_path = True
+            found_cl_path = SCons.Util.WhereIs('cl', v)
+            found_cl_envpath = SCons.Util.WhereIs('cl', env['ENV'].get(k, []))
         env.PrependENVPath(k, v, delete_existing=True)
         debug("env['ENV']['%s'] = %s", k, env['ENV'][k])
 
-    # final check to issue a warning if the compiler is not present
-    if not find_program_path(env, 'cl'):
-        debug("did not find %s", _CL_EXE_NAME)
+    debug("cl paths: d['PATH']=%s, ENV['PATH']=%s", repr(found_cl_path), repr(found_cl_envpath))
+
+    # final check to issue a warning if the requested compiler is not present
+    if not found_cl_path:
+        warn_msg = "Could not find requested MSVC compiler 'cl'."
         if CONFIG_CACHE:
-            propose = f"SCONS_CACHE_MSVC_CONFIG caching enabled, remove cache file {CONFIG_CACHE} if out of date."
+            warn_msg += f" SCONS_CACHE_MSVC_CONFIG caching enabled, remove cache file {CONFIG_CACHE} if out of date."
         else:
-            propose = "It may need to be installed separately with Visual Studio."
-        warn_msg = f"Could not find MSVC compiler 'cl'. {propose}"
+            warn_msg += " It may need to be installed separately with Visual Studio."
+        if found_cl_envpath:
+            warn_msg += " A 'cl' was found on the scons ENV path which may be erroneous."
```

Comment:
if the env['ENV']['PATH'] contains path to the requested and also otherwise found cl.exe, will this message get issued?

If the user specifically sets all the env['ENV'][*]'s set by msvc initialization will this get issued even if that is correct?

Reply:
> if the env['ENV']['PATH'] contains path to the requested and also otherwise found cl.exe, will this message get issued?

No.

```
    if not found_cl_path:   <--- ONLY IF CL.EXE NOT FOUND IN SETUP DICT
        warn_msg = "Could not find requested MSVC compiler 'cl'."
        if CONFIG_CACHE:
            warn_msg += f" SCONS_CACHE_MSVC_CONFIG caching enabled, remove cache file {CONFIG_CACHE} if out of date."
        else:
            warn_msg += " It may need to be installed separately with Visual Studio."
        if found_cl_envpath:
            warn_msg += " A 'cl' was found on the scons ENV path which may be erroneous."
        debug(warn_msg)
        SCons.Warnings.warn(SCons.Warnings.VisualCMissingWarning, warn_msg)
```

> If the user specifically sets all the env['ENV'][*]'s set by msvc initialization will this get issued even if that is correct?

I'm not exactly sure what you're asking.  The paths that come through this code are a request for the setup dictionary from MSVC_USE_SCRIPT (user script specification), MSVS_USE_SETTINGS (user dictionary specification), and MSVC autodetection.

Bypassing detection exits before this point:
```
    else:
        debug('MSVC_USE_SCRIPT set to False')
        warn_msg = "MSVC_USE_SCRIPT set to False, assuming environment " \
                   "set correctly."
        SCons.Warnings.warn(SCons.Warnings.VisualCMissingWarning, warn_msg)
        return None
```

The message is invoked when a requested setup dictionary does not contain a compiler executable.  The possibly erroneous fragment is indicating the there is a compiler in the user's path which may not be what was expected as the requested setup dictionary failed to include a compiler executable.

Reply:
OK. that's not clear from the CHANGES.txt (and release.txt) blurbs.
So these changes only apply when MSVC_USE_SETTINGS is set?

Reply:
MSVC_USE_SCRIPT => user-specified script (intent to get setup config from provided script output)
MSVC_USE_SETTINGS => user-specified dictionary (intent to get setup config from provided dictionary)
MSVC_VERSION => auto-detection (intent to get setup config from selected script output)
BYPASS => early-exit

Basically, the user requested a "valid" setup configuration.  Failing to find a compiler executable in the setup dictionary indicates that the request failed.

The augmented message is necessary because if/when there is a compiler executable already in the user's environment (which may or may not be the desired compiler), the build may continue and could possibly complete successfully with the "wrong" compiler.

There is no guarantee that the configured system environment is the same as the requested setup environment.

Reply:
Perhaps the terminology is confusing: at the code level the "setup config" is a dictionary.  All three methods that do not exit early return a setup configuration dictionary.

Reply:
So please be explicit about such in the CHANGES and RELEASE. While you may immediately know the correlation between the text and the variables above, most people won't.

Reply:
The correlation between the text and the variables above *did not change*.

With the exception of the bug fix, all that is different is that possibly another sentence is included.

* The bug fix (inspecting the os path via find_program_path which was removed).

   CHANGES.TXT: `The compiler detection no longer considers the host os environment path. `

   RELEASE.TXT: `The search for the msvc compiler executable (cl.exe) no longer inspects the OS system path in certain situations when setting up the msvc environment.`

* If necessary, the new third sentence.

  CHANGES.TXT: ` In addition, existence of the msvc compiler executable is checked in the detection dictionary and the scons ENV path before the detection dictionary is merged into the scons ENV.  Different warnings are produced when the msvc compiler is not detected in the detection dictionary based on whether or not an msvc compiler was detected in the scons ENV path (i.e., already exists in the user's ENV path prior to detection)` 

  RELEASE.TXT: `When the msvc compiler executable is not found during setup of the msvc environment, the warning message issued takes into account whether or not a possibly erroneous compiler executable was already present in the scons environment path.`

Reply:
Old message:
`Could not find MSVC compiler 'cl'. It may need to be installed separately with Visual Studio.`

New message (no cl on env path) with word `requested`:
 `Could not find requested MSVC compiler 'cl'. It may need to be installed separately with Visual Studio.`

New message (cl on env path) with word `requested` and more information:
 `Could not find requested MSVC compiler 'cl'. It may need to be installed separately with Visual Studio. A 'cl' was found on the scons ENV path which may be erroneous.`